TS-44 FEATURE - User Burger Menu, Profile & My Teamsheets Page

dcarg opened the pull request:
Few things still to figure out here but as I am editing a fair bit want to get this PR up, we need to think through some things then should try to get this ticket in without adding too much more scope to it.

User Burger Menu
---
Just went with a simple pop open menu but not sure if this should be full screen width on mobile or something?
![Screenshot 2024-01-27 at 1 44 59 pm](https://github.com/dcarg/teamsheet/assets/39669538/bed85018-be40-4b5c-a5a1-cbbfa7540554)

My TeamSheets Page
----
Thing that needs figuring out here is that whilst i can list the teamsheets, display their teamsheet and team names, I need to get access to the sport to be able to provide the edit and share urls to navigate to
Currently to get the sports would need to get the Teamsheets > Team > CompetitionTeam > Competition > Sport relation... Must be a better way? Thoughts?
![Screenshot 2024-01-27 at 1 44 50 pm](https://github.com/dcarg/teamsheet/assets/39669538/6ef4a805-fc2f-442b-9ea3-e97c03ed6e17)

Profile Page
----
Can Hide this item in the burger menu until we are ready but essentially I think we need to steup a useSetState and useForm hooks to manage the entityState then fire off mutation request. I will work on this but any thoughts?
Also whilst we can edit the field username as thats only a field on our db. If we want users to edit their email or name then we need to ensure we are also updating on Kinde?  My thinking here is mainly just I want users to be able to edit their username and perhaps we change it from their email to saving it as the first 5 letters of their email on create or something? But I want users to be able to edit their username so that when we share to socials we can pre fill their username

![Screenshot 2024-01-27 at 1 45 06 pm](https://github.com/dcarg/teamsheet/assets/39669538/985b5271-78e1-4493-81a5-4542d9951084)

dcg-domain committed:
User Burger Menu

dcg-domain committed:
UserMenu, Profile and My Team Sheets Pages

dcarg commented on src/app/shared/components/PageHeader/index.tsx:
Plan on adding an actionContent prop for action buttons to go up in the top right.

i.e. on the my team sheets page a create new button
on the share page here can go the duplicate and create new buttons etc

Mark-Ball commented on src/app/rugby/[team]/page.tsx:
```diff
@@ -64,6 +65,8 @@ const Page = async (props: PageProps) => {
 
   return (
     <TeamContent players={players} team={team} teamSheet={teamSheet}>
+      <TopBar />
```
this is still just an interim solution, right? the intention is still to replace this team selection with icons?

dcarg replied:
correct

Mark-Ball commented on src/app/shared/components/UserMenu/index.tsx:
```diff
@@ -0,0 +1,69 @@
+    <div className="relative">
+      <button
+        className="flex items-center text-white focus:outline-none"
+        onClick={toggleMenu}
```
i would just write the function in-line here instead of assigning it to a variable above, since we only use it this one time

Mark-Ball commented on src/app/shared/components/UserMenu/index.tsx:
```diff
@@ -0,0 +1,69 @@
+
+  return (
+    <div className="relative">
+      <button
```
can we use our new Button component here? now we have that i don't think we should ever have to write a html button

dcarg replied:
Updated

Mark-Ball commented on src/app/shared/components/UserMenu/index.tsx:
```diff
@@ -0,0 +1,69 @@
+
+      {isOpen && (
+        <div className="absolute right-0 mt-2 w-48 bg-white border rounded shadow-lg">
+          <div className="py-2 items-center justify-items-center justify-center content-center">
```
`justify-items-center` is for grid so we don't need it here

dcarg replied:
don't need any of them actually lol - when I chuck them all in I was clearly struggling haha

Mark-Ball commented on src/app/shared/components/UserMenu/UserMenuItem/index.tsx:
```diff
@@ -0,0 +1,33 @@
+interface UserMenuItemProps {
+  href?: string,
+  icon: IconDefinition,
+  onClick?: () => void,
```
do we have plans to use this with an onClick in the future? currently this never gets passed

dcarg replied:
yeh nah probably not - I had added this for the logout link assuming there would be a function from kinde to pass to an onClick.

Doesn't appear to be so you instead pass through the logoutLink component. 

I didn't love passing through either text or a reactnode to a text param but tailwind does style the logout link component text as well. 

Have updated with optional props:
text/: string,
textComponent?: React.ReactNode,

This component is still inside a div with tailwind classes but I think we are only every going to pass either text or specifically this logoutLink component lol... 

This'll do

Mark-Ball commented on src/app/shared/components/UserMenu/UserMenuItem/index.tsx:
```diff
@@ -0,0 +1,33 @@
+
+  return (
+    <a
+      className="flex px-4 py-2 hover:text-cyan-400"
```
the hover behaviour here is a little wonky, we need to set a class on the parent `group`, then connect that class to the hover on the children `group-hover`

```
<a
  className="flex px-4 py-2 group"
  href={href}
  onClick={onClick}
>
  <div className="flex justify-center w-[30px] group-hover:text-cyan-400">
    <FontAwesomeIcon icon={icon}/>
  </div>

  <div className="ml-2 font-semibold text-gray-800 group-hover:text-cyan-400">
    {text}
  </div>
</a>
```

dcarg replied:
Nice one! This is exactly what I was looking for

Mark-Ball commented on src/app/profile/_EditProfile/index.tsx:
i don't really understand our intention here
1.  how will we use the users `username`?
2. updating the email in our db doesn't seem right. if we want to do this i think we need to send a request to update on kinde, then when we get a successful response update in our db. But from a quick look around on kinde, it doesn't appear possible to update a user's email

Mark-Ball commented on src/app/shared/components/Button/index.tsx:
```diff
@@ -0,0 +1,23 @@
+  className?: string,
+  disabled?: boolean,
+  href?: string,
+  onClick?: Function,
```
changing this to `() => void` seems to make typescript happy

Mark-Ball commented on src/app/shared/components/Button/index.tsx:
```diff
@@ -0,0 +1,23 @@
+interface ButtonProps {
+  className?: string,
+  disabled?: boolean,
+  href?: string,
```
i don't think a button needs an href, if we need a link we will just use `Link` from next/link

also once you fix the type warning on onClick, it shows another one here saying the `href` is not available on button

Mark-Ball commented on src/app/shared/components/Button/index.tsx:
```diff
@@ -0,0 +1,23 @@
+  const { className, disabled, href, onClick, text } = props
+
+  return (
+    <button
```
may as well add `disabled={disabled}`

Mark-Ball commented on src/app/shared/components/FormField/index.tsx:
```diff
@@ -0,0 +1,24 @@
+  const { className, children, label, labelProps } = props
+
+  return (
+    <div className={clsx("flex flex-col ", className)}>
```
remove extra space

Mark-Ball commented on src/app/shared/components/Label/index.tsx:
```diff
@@ -1,12 +1,21 @@
 
   return (
-    <div className="mb-1 text-slate-500 text-sm tracking-wider">{children}</div>
+    <div className={clsx(
```
can we pls lint with className always on next line if we are breaking over multiple lines?
```
<div
  className={clsx(
    "mb-1 text-slate-500 text-sm tracking-wider",
    className,
  )}
>
  {children}
</div>
```

dcarg replied:
yep ok

Mark-Ball commented on src/app/shared/components/Line/index.tsx:
```diff
@@ -0,0 +1,15 @@
+const Line: React.FC<LineProps> = (props: LineProps) => {
+  const { className } = props
+
+  return <div className={clsx("h-[1px] mt-3 w-full bg-slate-200", className)}></div>
```
this div can just be self-closing

Mark-Ball commented on src/app/shared/components/Line/index.tsx:
```diff
@@ -0,0 +1,15 @@
+  className?: string;
+}
+
+const Line: React.FC<LineProps> = (props: LineProps) => {
```
i didn't really understand this, what is the reason this is typed as `React.FC` when none of our other functional components are?

dcarg replied:
Ah yes this was left over from a copy pasta - generated this component from chatGPT lol and forgot to change all of the code

Mark-Ball commented on src/app/teamSheets/page.tsx:
```diff
@@ -0,0 +1,55 @@
+  const kindeCurrentUser = await getUser()
+
+  const currentUser = await findOrCreateUser(kindeCurrentUser)
+  const { id: currentUserId } = currentUser
```
getting a ts warning here because `currentUser` can be null, easiest fix is to move the redirect up

```
const currentUser = await findOrCreateUser(kindeCurrentUser)
if (!currentUser) redirect('/#')
```

Mark-Ball commented on src/app/teamSheets/page.tsx:
```diff
@@ -0,0 +1,55 @@
+    },
+    include: {
+      team: true,
+    }
```
probably a good idea to `orderBy` something here... maybe title or updatedAt

Mark-Ball commented on src/app/teamSheets/_TeamSheetListItem/index.tsx:
```diff
@@ -0,0 +1,56 @@
+  const { editId, shareId, team, title } = teamSheet
+  const { title: teamTitle } = team
+
+  const pathname = usePathname()
```
pathname won't help us here because we are not adding to the `/teamsheets` route, so this can be removed and this can be a server component

Mark-Ball commented on src/app/teamSheets/page.tsx:
```diff
@@ -0,0 +1,55 @@
+    where: {
+      userId: currentUserId,
+    },
+    include: {
```
regarding your comment in the other component, i don't see how to do this other than including all those nested entities in the teamSheet request. the only optimisation i can think of it that all the competitions the team is in will be for the same sport, so is it possible to only request the first competitionTeam record, instead of all?

```
include: {
  team: {
    include: {
      competitionTeams: {
        include: {
          competition: {
            include: {
              sport: true
            },
          },
        },
      },
    },
  },
},
```

dcarg replied:
I want to get all competitionTeams to see if the team has multiple teams or not. If they don't have multiple competitions then will display team name and competition name on the team sheet list item

dcarg committed:
Fixes to userMenu, userMenuItem and some shared component styling/props

dcarg committed:
vertically center userMenu text

dcarg committed:
add is-mobile package to conditionally render user given name when not mobile

dcarg committed:
add orderBy to TeamSheets query

dcarg commented:
Cleaned up added to some of those shared components.

Added is-mobile package for rendering a users given name next to the user burger menu if not mobile

commented out profile page for now

Mark-Ball commented on src/app/teamSheets/_TeamSheetListItem/index.tsx:
```diff
@@ -0,0 +1,67 @@
+          <div>{teamTitle}</div>
+
+          {!hasMultipleCompetitionTeams && (
+            <div>, {competitionTitle}</div>
```
i don't think showing this makes sense

there is currently no way of knowing which competition a teamSheet is for, we would have to add a relation for competition directly on teamSheet for that

do we expect a user to pick different players for the same team in different competitions?

dcarg replied:
so Yep summarising our discussion today.

If a team is in multiple competitions. I.e the Wallabies in 'International' & 'The Rugby Championship' then we will not display the competition title

A User cannot pick a different set of players for a team for each competition. You can make multiple wallabies teamSheets. but you dont link them to a particular competition. The wallabies team can just be selected in multiple competitions to then create a teamSheet for.

If a team is only in one competition. I.e. the NSW Waratahs are only in the Super Rugby Pacific Competition. Which will be the case for most teams that are not Interntational teams, then we can display the competition name just for more information.

dcarg committed:
Merge branch 'main' into feature/TS-44

dcarg commented:
I merged your duplicate team sheet then merged master into this branch.

You'd taken my Button component, added it and improved on it. I'd made some other changes in this PR. 

I've opted for the text prop over children atm not sure which is better tbh but think this'll do for now and can update/improve on it later.

dcarg committed:
Merge branch 'main' into feature/TS-44

dcarg committed:
Merge branch 'main' into feature/TS-44

dcarg committed:
update hasMultipleCompetitionTeams

dcarg committed:
fix button, add empty teamsheet title

dcarg commented:
Fixed the buttons (still need to style them all better and will do but lets get this in)

Also added a default teamsheet title to display in the even a logged in user hasn't named a teamsheet. Maybe we should set a default like user's given name/username team name i.e. Mark's Wallabies

![Screenshot 2024-02-06 at 19 14 13](https://github.com/dcarg/teamsheet/assets/39669538/9a42c58a-6ca2-428a-a828-d3332157dcfd)

Mark-Ball committed:
lint Button component and fix Duplicate button appearance
